_posts/2025-03-20-phased-strimzi-upgrade-example.md

```diff
@@ -0,0 +1,637 @@
+---
+layout: post
+title: "Phased upgrades of Strimzi managed Kafka fleets"
+date: 2025-03-20
+author: tom_cooper
+---
+
+## TL;DR
+
+When you upgrade an instance of the Strimzi Cluster Operator you will force **all** the Kafka clusters that it manages to roll. 
+For large fleets this could be quite disruptive.
+It would be nice to be able to decide which Kafka clusters upgrade to the new Strimzi version and when.
+
+This blog provides an example of how phased Strimzi upgrades can be performed.
+By using kustomize to create a deployment that allows multiple Strimzi versions to run concurrently and the `STRIMZI_CUSTOM_RESOURCE_SELECTOR` to allow us to specify which Strimzi instance manages which Kafka cluster.
+
+An [example repository](https://github.com/tomncooper/strimzi-phased-upgrade) is provided showing the final setup and providing example scripting and templates for adding new Strimzi versions.
+
+## Background
+
+Strimzi handles the [upgrade of Kafka clusters](https://strimzi.io/docs/operators/latest/deploying#proc-upgrade-kafka-kraft-str) for you, making it simple and easy for infrastructure teams to handle these complex operations. 
+The [upgrade of the Strimzi operator](https://strimzi.io/docs/operators/latest/deploying#assembly-upgrade-cluster-operator-str) itself is also straight-forward. 
+However, when you have a large fleet of Kafka clusters being managed by Strimzi, there are several elements you must keep in mind when performing the upgrade. 
+
+First, lets go over what a Strimzi upgrade involves:
+
+### CRDs
+
+The first step in a Strimzi operator upgrade is the Custom Resource Definitions (CRDs) for the Strimzi Custom Resources (CRs): `Kafka`, `KafkaTopic`, `KafkaUser` etc. 
+These are Kubernetes cluster wide resources and need to match the most recent version of Strimzi you have installed on the Kubernetes Cluster. 
+They need to match the most recent version as new fields could have been added to the CRDs that the new Strimzi version expects to be there. 
+Strimzi CRDs are backwards compatible (within the same API version), so older Strimzi versions are ok with newer CRDs (they will ignore the new fields) but not the other way around.
+
+### Strimzi Deployment
```

Comment:
Should you also mention the cluster roles and cluster role bindings? They will also conflict between parallel operator installations so it should be worth mentioning them.

Reply:
So this section was supposed to be about what you need to do for a normal Stimzi upgrade and the consequences of that, basically setting up why you might want to have multiple strimzi versions installed.

The ClusterRole/Roles and their bindings are only an issue once you do that and we haven't proposed that yet at this point in the post.

Reply:
Yet you talk about these things with CRDs right above this, or?